lib/ddtrace/correlation.rb

```diff
@@ -6,35 +6,109 @@ module Datadog
   # Contains behavior for managing correlations with tracing
   # e.g. Retrieve a correlation to the current trace for logging, etc.
   module Correlation
-    # Struct representing correlation
-    Identifier = Struct.new(:trace_id, :span_id, :env, :service, :version) do
-      def initialize(*args)
-        super
-        self.trace_id = trace_id || 0
-        self.span_id = span_id || 0
-        self.env = env || Datadog.configuration.env
-        self.service = service || Datadog.configuration.service
-        self.version = version || Datadog.configuration.version
+    # Represents current trace state with key identifiers
+    class Identifier
+      attr_reader \
+        :env,
+        :service,
+        :span_id,
+        :span_name,
+        :span_resource,
+        :span_service,
+        :span_type,
+        :trace_hostname,
+        :trace_id,
+        :trace_name,
+        :trace_origin,
+        :trace_process_id,
+        :trace_resource,
+        :trace_runtime_id,
+        :trace_sampling_priority,
+        :trace_service,
+        :version
+
+      # rubocop:disable Metrics/CyclomaticComplexity
+      # rubocop:disable Metrics/PerceivedComplexity
+      def initialize(
+        env: nil,
+        service: nil,
+        span_id: nil,
+        span_name: nil,
+        span_resource: nil,
+        span_service: nil,
+        span_type: nil,
+        trace_hostname: nil,
+        trace_id: nil,
+        trace_name: nil,
+        trace_origin: nil,
+        trace_process_id: nil,
+        trace_resource: nil,
+        trace_runtime_id: nil,
+        trace_sampling_priority: nil,
+        trace_service: nil,
+        version: nil
+      )
+        @env = env || Datadog.configuration.env
+        @service = service || Datadog.configuration.service
+        @span_id = span_id || 0
+        @span_name = span_name && span_name.dup.freeze
+        @span_resource = span_resource && span_resource.dup.freeze
+        @span_service = span_service && span_service.dup.freeze
+        @span_type = span_type && span_type.dup.freeze
+        @trace_hostname = trace_hostname && trace_hostname.dup.freeze
+        @trace_id = trace_id || 0
+        @trace_name = trace_name && trace_name.dup.freeze
+        @trace_origin = trace_origin && trace_origin.dup.freeze
+        @trace_process_id = trace_process_id
+        @trace_resource = trace_resource && trace_resource.dup.freeze
+        @trace_runtime_id = trace_runtime_id && trace_runtime_id.dup.freeze
+        @trace_sampling_priority = trace_sampling_priority
+        @trace_service = trace_service && trace_service.dup.freeze
+        @version = version || Datadog.configuration.version
+
+        # Dup and freeze strings so they aren't modified by reference.
```

Comment:
I believe the general comment about "dup+freeze" applies to the variables above as well.

I understand that most of the variables above do not originate from `Datadog.configuration`, but I'm just describing the "dup+freeze" intent in this initializer.

Reply:
Yes, this is true. The comment was meant to highlight why these were done below, after the rest were already set (so as to simplify the logic a bit.)